src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -63,14 +63,100 @@ internal static IEnumerable<RuntimeLibrary> GetCandidateLibraries(DependencyCont
                 return Enumerable.Empty<RuntimeLibrary>();
             }
 
-            return dependencyContext.RuntimeLibraries.Where(IsCandidateLibrary);
+            var candidatesResolver = new CandidatesResolver(dependencyContext.RuntimeLibraries, ReferenceAssemblies);
+            return candidatesResolver.GetCandidates();
         }
 
-        private static bool IsCandidateLibrary(RuntimeLibrary library)
+        private class CandidatesResolver
         {
-            Debug.Assert(ReferenceAssemblies != null);
-            return !ReferenceAssemblies.Contains(library.Name) &&
-                library.Dependencies.Any(dependency => ReferenceAssemblies.Contains(dependency.Name));
+            private readonly IDictionary<string, Dependency> _dependencies;
+
+            public CandidatesResolver(IReadOnlyList<RuntimeLibrary> dependencies, ISet<string> referenceAssemblies)
+            {
+                _dependencies = dependencies
```

Comment:
The list is not enough, I need to keep track of who is a candidate, who is not, and who is unknown, otherwise I would  visit nodes on the tree multiple times (dependencies contains a list of all sub-trees) which has way higher algorithmic complexity. Using a dictionary allows me to efficiently find the given RuntimeLibrary based on a Dependency O(n) vs O(1) for finding an element. And In my opinion it makes it clearer. Am I missing something?